### PR TITLE
Segmentation module

### PR DESCRIPTION
**Pull request by NikitaS4**

### CODE REVIEW

File: SegmentationModule.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            batch_size: the number of samples in a single batch to fit"""
+        assert len(x_train.shape) == 3, "X_train shape must be (samples, time, embeddings)"
+        assert len(x_valid.shape) == 3, "X_valid shape must be (samples, time, embeddings)"
+        assert len(y_train.shape) == 3, "Y_train shape must be (samples, time, 1)"
+        assert len(y_valid.shape) == 3, "Y_valid shape must be (samples, time, 1)"
+        assert isinstance(checkpoint_file, str) and checkpoint_file.endswith(".h5"), ("Checkpoint file must be string"
+                                                                                      "ends with .h5")
+        if self.model is None:  # build default model
+            self.__build_new_model()
+        y_tr_crf, y_val_crf = Models.data_to_crf(y_train, y_valid)  # convert to CRF-suitable format
```
Comment (OlegBEZb) [inline SegmentationModule.py:38]: name method `convert_to_crf_format` or like that and there will be no need of comment)

File: SegmSubmodules/Evalutation.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+def draw_mask_plots(prediction, ground_truth, plot_file):
+    """Draws plots of the predicted and ground truth mask and saves it to plot_file as image
+
+    Args:
+        prediction: the prediction binary mask vector
+        ground_truth: the ground truth binary mask vector
+        plot_file: path to file where plot will be saved"""
+    # Have to add captions, tags and labels
+    plt.figure()
+    plt.plot([x for x in range(len(prediction))], prediction, [x for x in range(len(prediction))], ground_truth)
```
Comment (OlegBEZb) [inline SegmSubmodules/Evalutation.py:18]: don't repeat yourself

File: SegmSubmodules/Evalutation.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    plt.savefig(plot_file)
+
+
+def count_metrics_on_sample(prediction, ground_truth, json_file):
```
Comment (OlegBEZb) [inline SegmSubmodules/Evalutation.py:22]: `metrics_list` parameter can be added. if empty, use a predefined list of sklearn metrics.

File: SegmSubmodules/Evalutation.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    """This class can be used to evaluate model. Here are methods with hardcoded labels for some data"""
+    @staticmethod
+    def get_epica():
+        """Get ground truth mask for Epica video
+
+        Returns:
+            the ground truth mask for Epica video"""
+        ratio = 100 / 96  # as 100 VGG embeddings are 96 seconds
+        q_len = 1112  # length of the sample
+        true_value = np.zeros(q_len, dtype=np.float32)
```
Comment (OlegBEZb) [inline SegmSubmodules/Evalutation.py:115]: please, rename the variable to sth like `gt_mask` or `ground_truth_mask` etc

File: SegmSubmodules/Evalutation.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    Returns:
+        scalar (intersection over union metrics value)"""
+    intersection = true * pred
+    true_inv = 1 - true
+    union = true + (true_inv * pred)
+
+    return K.sum(intersection) / K.sum(union)
+
+
+class GroundTruthHardcoder:
```
Comment (OlegBEZb) [inline SegmSubmodules/Evalutation.py:105]: discussable: probably, this is not the correct place for such class

File: SegmSubmodules/Models.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    Args:
+        path: path to the model checkpoint
+
+    Returns:
+        loaded keras model"""
+    model = load_model(path, custom_objects=__create_custom_objects())
+    return model
+
+
+def predict_mask_long_sample(x_data, crf_model):
```
Comment (OlegBEZb) [inline SegmSubmodules/Models.py:48]: You can simplify the name of the function via adding an assert for the length of the sample

File: SegmSubmodules/Models.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    single_dim_sample = x_data.shape[1]  # total length
+    tracks = int(single_dim_sample / SEQ_LEN)  # count the number of pieces to predict on
+    sample_mask = np.zeros((single_dim_sample, 1), dtype=np.float32)  # init tensor for the mask
+
+    for i in range(tracks):  # make prediction for each piece
+        q = crf_model.predict(x_data[:, i * SEQ_LEN:(i + 1) * SEQ_LEN, :])
+        q = q[0, :, 0]
+        sample_mask[i * SEQ_LEN:(i + 1) * SEQ_LEN, 0] = q  # add to the sample mask
+
+    q = crf_model.predict(x_data[:, tracks * SEQ_LEN:, :])  # predict on the last piece
```
Comment (OlegBEZb) [inline SegmSubmodules/Models.py:68]: Please, check out the border case. If the length of the input_sample is multiple of SEQ_LEN, out of range can happen. Correct me if I'm wrong, please.

File: SegmSubmodules/Models.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    q = q[0, :, 0]
+    sample_mask[tracks * SEQ_LEN:, 0] = q
+    sample_mask = sample_mask.reshape(-1)  # reshape mask into binary vector
+    return sample_mask
+
+
+def data_to_crf(y_tr, y_val):
+    """Translates label tensors from data generator format to CRF-supportable format
```
Comment (OlegBEZb) [inline SegmSubmodules/Models.py:76]: don't repeat yourself

File: SegmSubmodules/PredictionCutter.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        stored as string in format 'hh:mm:ss'
+
+        extension: target extension (e. g., ".wav" for sound)"""
+    i = 1  # counter for video piece
+    for interval in prediction_intervals:
+        timedelta_interval = __time_to_seconds(interval)  # convert to timedelta interval
+        begin = timedelta_interval[0].total_seconds()  # convert to total seconds as ffmpeg gets seconds
+        end = timedelta_interval[1].total_seconds()
+        duration = end - begin
+        piece_name = target_path + "_piece_" + str(i) + extension  # create name of the file to save piece
```
Comment (OlegBEZb) [inline SegmSubmodules/PredictionCutter.py:24]: https://stackoverflow.com/questions/13944387/why-use-os-path-join-over-string-concatenation

File: SegmSubmodules/PredictionCutter.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import datetime
+import subprocess
+
+
+def cut_file(path_to_file, target_path, prediction_intervals, extension=".wav"):
```
Comment (OlegBEZb) [inline SegmSubmodules/PredictionCutter.py:5]: slice maybe?

File: SegmentationModule.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        Args:
+             x_data: tensor of embeddings to make prediction
+
+        Returns:
+            the list of time intervals where time interval is a list of the beginning time and the end time,
+            time stored in strings in format: 'hh:mm:ss'
+        """
+        assert self.model is not None, "Can't predict, model has not been loaded yet"
+        assert len(x_data.shape) == 3, "X_data shape must be (samples, duration, embeddings)"
+        # self.model must not be None
```
Comment (OlegBEZb) [inline SegmentationModule.py:65]: repetition  of line 63

File: SegmentationModule.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+        sample_mask = self.__get_raw_prediction(x_data)  # firstly, get raw prediction
+        smooth_mask = self.__get_smooth_mask(sample_mask)  # smooth mask
```
Comment (OlegBEZb) [inline SegmentationModule.py:67]: add a flag for smoothing. or maybe create a possibility to change smoothing techniques

File: SegmentationModule.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+            the list of time intervals where time interval is the list of the beginning time and the end time. Time
+            stored as string of format 'hh:mm:ss' """
+        sec_to_time = lambda arg: str(datetime.timedelta(seconds=arg))  # callback to translate seconds to string
+        vgg_rescale = 0.96  # 100 embeddings of VGG is 96 seconds
+        # translate embedding-time intervals to time-time intervals
+        intervals = [[sec_to_time(int(vgg_rescale * q)) for q in v] for v in abs_intervals]
+        return intervals
+
+    @staticmethod
+    def __define_callback_list(checkpoint_file):
```
Comment (OlegBEZb) [inline SegmentationModule.py:220]: I suppose we should have the possibility to use custom callback_list. But for the purpose of the undefined parameter, this default callback_list can be used

File: Pipeline/Segmentation/SegmentationModule.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+        fragments_list = []
+        start_segment = idxs[0]
+        finish_segment = idxs[0]
+        prev_val = idxs[0]
+        for v in idxs:  # unite neighbor mask ones into intervals
+            if v - prev_val < max_lag:
+                prev_val = v
+                finish_segment = v
+            else:
+                if finish_segment - start_segment > min_music_len:
```
Comment (mitenevav) [inline Pipeline/Segmentation/SegmentationModule.py:182]: maybe make one line in the "if" section and delete the "else" section

File: Pipeline/Segmentation/SegmSubmodules/Evaluation.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
+import numpy as np
+import json
+import matplotlib.pyplot as plt
+from sklearn.metrics import log_loss, f1_score, r2_score, confusion_matrix, classification_report, roc_curve, auc, \
+    roc_auc_score, precision_score, recall_score
+import keras.backend as K
+
+
+def draw_mask_plots(prediction, ground_truth, plot_file):
```
Comment (mitenevav) [inline Pipeline/Segmentation/SegmSubmodules/Evaluation.py:9]: the legend is missing

**Pull request merged**